### ISSUE TITLE
metadata headers: numbers in field names not read; non-string fields not converted

### ISSUE THREAD

**Issue by vchudnov-g** [googleapis/gapic-generator-go#514]
As part of https://github.com/googleapis/gapic-showcase/pull/532, I am adding an RPC with http rules for path params:

```
  rpc RepeatDataSimplePath(RepeatRequest) returns (RepeatResponse) {
    option (google.api.http) = {
      get: "/v1beta1/repeat/{info.f_string}/{info.f_int32}/{info.f_double}/{info.f_bool}:simplepath"
    };
```

which causes gapic-generator-go to generate the following code (formatted for readability here):
```
func (c *ComplianceClient) RepeatDataSimplePath(ctx context.Context, req *genprotopb.RepeatRequest, opts ...gax.CallOption) (*genprotopb.RepeatResponse, error) {
	md := metadata.Pairs("x-goog-request-params", 
                    fmt.Sprintf("%s=%v&%s=%v&%s=%v&%s=%v", 
                      "info.f_string", url.QueryEscape(req.GetInfo().GetFString()), 
                      "info.f_int", url.QueryEscape(req.GetInfo().GetFInt()), 
                      "info.f_double",  url.QueryEscape(req.GetInfo().GetFDouble()), 
                      "info.f_bool", url.QueryEscape(req.GetInfo().GetFBool())))
…
}
```

This client leads to the following compile-time errors:

- [x] `req.GetInfo().GetFInt undefined (type *genproto.ComplianceData has no field or method GetFInt)`
  - There are fields `f_int32` and `f_int64`, but no field `f_int`. I suspect we're only expecting/parsing letters in the field params
- [x] `cannot use req.GetInfo().GetFDouble() (type float64) as type string in argument to url.QueryEscape`
   - I suspect this has only been tested on path params for string fields in the request
- [x] `cannot use req.GetInfo().GetFBool() (type bool) as type string in argument to url.QueryEscape`
   - ditto

For the time being, I am working around this by commenting out the offending lines via Showcases's `compile_protos.go` utility.

**Comment by noahdietz** [comment]
For the first one it is definitely a regexp issue. Only capturing letters, not numbers. Easily fixed.

**Comment by noahdietz** [comment]
For the other two, we can add a check if the "field type != string" and if not, we can either:
A. generate a call to `strconv.Format(Bool|Float|Int)` around the `req.GetInfo().GetF{Double|Int64|Bool}()`
B. exclude the wrapping call to `url.QueryEscape`, since a formatted number or bool should always be valid

[B] seems a lot easier and cleaner. WDYT?

**Comment by vchudnov-g** [comment]
With either alternative, we should support all [proto scalar types](https://developers.google.com/protocol-buffers/docs/proto#scalar) (see the mentioned PR for a proto that uses them all). In particular, fields of type `bytes` will also need some sort of encoding.

As long as all the characters can be encoded in a way they can be decoded, [B] seems fine:
- Bools only have letters, so that's fine. 
- Int types only have digits, so that's fine. 
- Float types could have periods (which should be fine) and exponents "e-" and "e+". I think we're already using "-" in headers, so that should be fine, but we need to check that "+" works as expected (note: URI-**de**coding would treat "+" as a space, so if the server reading the servers is doing that, we should properly URI encode the "+")

Also worth syncing with @software-dov since this field decoding/encoding might overlap with some of his REST work---not sure how far along he is with that.

**Comment by noahdietz** [comment]
For float types, maybe we have to use [A] so that we can control the use of `e-`/`e+` in the formatting. We could have it only use decimals. See [strconv.FormatFloat](https://golang.org/pkg/strconv/#FormatFloat). Maybe we just use mode `'f'` (-ddd.dddd, no exponent). Here is a quick example: https://play.golang.org/p/lrHVcmCspAo

For bytes, I'm open to ideas...is there an existing usage to reference (other than a showcase test)?

**Comment by noahdietz** [comment]
We can use the `%f` format specifier will use the `'f` format mode (so decimal, with no exponent). So we could `fmt.Printf("%f", 0.000004)` and it would produce `0.000004`, not `4.000000e-06` (that would be the `'e'` mode). The default specifier `%v` would use `'b'` mode (decimalless scientific notation with exponent a power of two) so we can't use that.

For bytes I'm not really sure what to do. It becomes a `[]byte` in Go, which we could easily convert to `string` and `QueryEscape` but idk if that would mangle the data. A use case would really help.

**Comment by vchudnov-g** [comment]
I'm not keen on skipping the exponent. Some numbers could be really large or really small and we don't want all those placeholder zeroes if we can skip them.

As for `bytes`, no, I don't have a real world use case. My mental model is that to zeroth order, converting to string ought to be lossless...except that strings are sequences of runes, so for all I know, the encoding of a rune may be different than the encoding for the bytes that represent it. We'd have to dig around---I would hope there is some library/answer that addresses this.

**Comment by noahdietz** [comment]
Ok, I hear you! `%g` could be better. But one point of clarification I'd like to make for discussion is that end users will not actually *see* the values here - the formatter is for runtime inclusion of the value from the message field accessor (i.e. `req.GetFDouble()`). Now, if there is a client trace or there are logs on the backend, it could show up as you say, with lots of zeros, which I'd understand being suboptimal and wanting to format it as an exponent for readability. I'm just concerned with encoding the `'e'` mode (with a `+`) and making the value more unreadable. 

As far as  precision, do you mean in the context of the `FormatFloat` argument? Or something else? If the former, using `-1` for the `prec` arg means it'll use the minimum required precision when formatting e.g. `000.000400` is printed as `0.0004`. LMK if I'm off.

For `bytes` agreed, lets do some digging!

**Comment by vchudnov-g** [comment]
For the precision, using -1 in `FormatFloat` sounds like the right thing, and it sounds like that corresponds to just `%g` in `fmt.*rintf` (`%g it is the smallest number of digits necessary to identify the value uniquely.`, [src](https://golang.org/pkg/fmt/))

Another thought: I have the vague impression that some systems/languages omit the `+` in `e+`. So I wonder whether we could do that before we stick the value in the header. I guess at the very least I'd want to check Go can parse values without the `+`

For `bytes`, yes, let's investigate.

To clarify: aside from the workaround (commenting out the lines) so the Showcase generated code will compile, I am not inspecting or otherwise using these metadata headers. This is not blocking my work; it just seems like a bug we should eventually resolve so the generator-go does the right thing in general.

**Comment by noahdietz** [comment]
Generated code to format a float, then remove the `+` then include it in the headers would be something like the following:

```go
func (c *ComplianceClient) RepeatDataSimplePath(ctx context.Context, req *genprotopb.RepeatRequest, opts ...gax.CallOption) (*genprotopb.RepeatResponse, error) {
	md := metadata.Pairs("x-goog-request-params", 
                    fmt.Sprintf("%s=%v&%s=%v&%s=%v&%s=%v", 
                      "info.f_string", url.QueryEscape(req.GetInfo().GetFString()), 
                      "info.f_double",  strings.Replace(fmt.Sprintf("%g", req.GetInfo().GetFDouble()), "+",  ""), 
…
}
```

We'd still use `%v` in the key-value mapping for the header because the float/double value has been converted to a string. 

To me, this seems a little like overkill, especially because the value will not always have a `+`. Using `%f` feels like it would be sufficient.

**Comment by noahdietz** [comment]
Released in `v0.17.2`. This _does not_ include a fix for `bytes` fields.
